core/services/cron/cron.go

```diff
@@ -0,0 +1,109 @@
+package cron
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/robfig/cron"
+	cronParser "github.com/robfig/cron/v3"
+
+	"github.com/smartcontractkit/chainlink/core/gracefulpanic"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+)
+
+// CronJob runs a cron jobSpec from a CronSpec
+type CronJob struct {
+	jobID int32
+
+	logger   *logger.Logger
+	runner   pipeline.Runner
+	Schedule string
+
+	pipelineSpec pipeline.Spec
+}
+
+func NewCronJob(jobID int32, logger *logger.Logger, schedule string, runner pipeline.Runner, spec pipeline.Spec) (*CronJob, error) {
+	return &CronJob{
+		jobID:        jobID,
+		logger:       logger,
+		Schedule:     schedule,
+		runner:       runner,
+		pipelineSpec: spec,
+	}, nil
+}
+
+// NewFromJobSpec() - instantiates a CronJob singleton to execute the given job pipelineSpec
+func NewFromJobSpec(
+	jobSpec job.Job,
+	runner pipeline.Runner,
+) (*CronJob, error) {
+	cronSpec := jobSpec.CronSpec
+	spec := jobSpec.PipelineSpec
+
+	if err := validateCronJobSpec(*cronSpec); err != nil {
+		return nil, err
+	}
+
+	cronLogger := logger.CreateLogger(
+		logger.Default.With(
+			"jobID", jobSpec.ID,
+			"schedule", cronSpec.CronSchedule,
+		),
+	)
+
+	return NewCronJob(jobSpec.ID, cronLogger, cronSpec.CronSchedule, runner, *spec)
+}
+
+// validateCronJobSpec() - validates the cron job spec and included fields
+func validateCronJobSpec(cronSpec job.CronSpec) error {
+	if cronSpec.CronSchedule == "" {
+		return fmt.Errorf("schedule must not be empty")
+	}
+	_, err := cron.Parse(cronSpec.CronSchedule)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Start implements the job.Service interface.
+func (cron *CronJob) Start() error {
+	cron.logger.Debug("Starting cron jobSpec")
+
+	go gracefulpanic.WrapRecover(func() {
```

Comment:
Why suppress panics?

Also, let's make starting/stopping synchronous.

Reply:
Yeah we need a way ensure this goroutine finishes when Close is called,

Reply:
roger, would it be a good case for a channel so when `.Close()` is called we can return from within the `.Start()` or `.run()` context?

Reply:
I would do something like this:
```
package cron

import (
	"context"
	"errors"
	"fmt"
	cronlib "github.com/robfig/cron"

	"github.com/smartcontractkit/chainlink/core/gracefulpanic"
	"github.com/smartcontractkit/chainlink/core/logger"
	"github.com/smartcontractkit/chainlink/core/services/job"
	"github.com/smartcontractkit/chainlink/core/services/pipeline"
)

// Cron runs a cron jobSpec from a CronSpec
type Cron struct {
	logger   *logger.Logger
	cronRunner *cronlib.Cron
	ctx context.Context
	cancel context.CancelFunc
	done chan struct{}
}

// NewFromJobSpec() - instantiates a Cron singleton to execute the given job pipelineSpec
func NewCron(
	jb job.Job,
	pr pipeline.Runner,
) (*Cron, error) {
	if jb.CronSpec == nil {
		return nil, errors.New("job must have cron spec present")
	}
	cronSpec := jb.CronSpec

	if cronSpec.CronSchedule == "" {
		return nil, fmt.Errorf("schedule must not be empty")
	}
	_, err := cronlib.Parse(cronSpec.CronSchedule)
	if err != nil {
		return nil, err
	}

	cronLogger := logger.CreateLogger(
		logger.Default.With(
			"jobID", jb.ID,
			"schedule", cronSpec.CronSchedule,
		),
	)
	ctx, cancel := context.WithCancel(context.Background())
	cronRunner := cronlib.New()
	err = cronRunner.AddFunc(cronSpec.CronSchedule, func() {
		_, _, err := pr.ExecuteAndInsertNewRun(ctx, *jb.PipelineSpec, pipeline.JSONSerializable{}, *cronLogger, true)
		if err != nil {
			if ctx.Err() != nil {
				cronLogger.Warnf("canceled cron run for jobID %v", jb.ID)	
				return
			}
			cronLogger.Errorf("error executing new run for jobID %v", jb.ID)
		}
	})
	if err != nil {
		return nil, err
	}

	return &Cron{
		logger:       cronLogger,
		cronRunner: cronRunner,
		ctx: ctx,
		cancel: cancel,
		done: make(chan struct{}),
	}, nil
}

// Start implements the job.Service interface.
func (c *Cron) Start() error {
	c.logger.Debug("Cron: Starting")
	go gracefulpanic.WrapRecover(func() {
		defer close(c.done)
		c.cronRunner.Run()
	})
	return nil
}

// Close implements the job.Service interface. It stops this instance from
// polling, cleaning up resources.
func (c *Cron) Close() error {
	c.logger.Debug("Cron: Closing")
	c.cancel() // Cancel any inflight cron runs.
	c.cronRunner.Stop()
	<-c.done
	return nil
}
```
noticed a few more things while writing that:
- We definitely want to use the same version of the cron lib throughout, looks like you currently have 2 different ones
- Doesn't look like there is a way to control the goroutines spawned by the libraries Run, so canceling the context is the best we can do
- Can simplify a fair amount in terms of functions and state on the object

Reply:
awesome thanks for this, very helpful, 

- reducing usage to use the v3 package, makes sense and it looks to be the lib we are using elsewhere
- cancel makes sense for the cron context, adding that to the `Close()`
- simplifying some of the interfaces as suggested, makes sense / not needed (ie `NewCron()`)

Reply:
also does it makes sense to remove the `	go gracefulpanic.WrapRecover(func() {` ? Thought we wouldn't want the node to crash if a cron job failed erroneously?

Reply:
I think its fine to leave that in - we're basically thinly wrapping some 3rd party lib which could panic I suppose